src/status/condition/mod.rs

```diff
@@ -0,0 +1,471 @@
+/// let new_status = ClusterStatus {
+///     conditions: compute_conditions(&old_status,
+///         &[
+///             &daemonset_condition_builder as &dyn ConditionBuilder,
+///             &statefulset_condition_builder as &dyn ConditionBuilder
+///         ]
+///     )
+/// };
+///
+/// ```
+pub fn compute_conditions<T: HasStatusCondition>(
+    resource: &T,
+    condition_builders: &[&dyn ConditionBuilder],
+) -> Vec<ClusterCondition> {
+    let mut new_resource_conditions = ClusterConditionSet::new();
+    for cb in condition_builders {
+        let conditions: ClusterConditionSet = cb.build_conditions();
+        new_resource_conditions = new_resource_conditions.merge(conditions, update_message);
+    }
+
+    let old_resource_conditions: ClusterConditionSet = resource.conditions().into();
+
+    old_resource_conditions
+        .merge(new_resource_conditions, update_timestamps)
+        .into()
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ClusterCondition {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// Last time the condition transitioned from one status to another.
+    pub last_transition_time: Option<Time>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// The last time this condition was updated.
+    pub last_update_time: Option<Time>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// A human readable message indicating details about the transition.
+    pub message: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// The reason for the condition's last transition.
+    pub reason: Option<String>,
+    /// Status of the condition, one of True, False, Unknown.
+    pub status: ClusterConditionStatus,
+    /// Type of deployment condition.
+    #[serde(rename = "type")]
+    pub type_: ClusterConditionType,
+}
+
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    Deserialize,
+    EnumCount,
+    Eq,
+    Hash,
+    JsonSchema,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    Serialize,
+)]
+#[serde(rename_all = "PascalCase")]
+pub enum ClusterConditionType {
+    #[default]
+    /// Available indicates that the binary maintained by the operator (eg: zookeeper for the
+    /// zookeeper-operator), is functional and available in the cluster.
+    Available,
+    /// Degraded indicates that the operand is not functioning completely. An example of a degraded
+    /// state would be if there should be 5 copies of the operand running but only 4 are running.
+    /// It may still be available, but it is degraded.
+    Degraded,
+    /// Progressing indicates that the operator is actively making changes to the binary maintained
+    /// by the operator (eg: zookeeper for the zookeeper-operator).
+    Progressing,
+    /// Paused indicates that the operator is not reconciling the cluster. This may be used for
+    /// debugging or operator updating.
+    Paused,
```

Comment:
We maybe want to call this `ReconciliationPaused` to make it more descriptive as discussed in arch meeting yesterday.

Reply:
Do you think it will be possible to also pause additional things in the future ?

Reply:
It can be further defined in the actual message as well. Im fine with both.

Reply:
> Do you think it will be possible to also pause additional things in the future ?

Maybe some product-specific stuff e.g. Airflow stopping scheduling stuff or an BackupDfs or SyncDfs object that periodically backups an hdfs to some other location. Here pause could express pausing the backup job and not the reconciliation. Just some random ideas. I would not like to impose my own opinion, happy to have a quick poll in Slack or similar

Reply:
I discussed with @razvan. We do not want (for now) to introduce more custom status fields. You have a valid point with the airflow example, but this should probably be reflected in an extra "backup" CRD status and forwarded to the "available" or "degraded" status that are already offered (with a custom message coming from the "backup" status). 
This is not against renaming paused to ReconcilationPaused (we do not have a strong opinion here).